### ISSUE TITLE
Allow viewing gain by showing the difference between valuation and cost basis

### ISSUE THREAD

**Issue by chvp** [simonmichael/hledger#1426]
As discussed on IRC.

A small example journal:
```
2020-01-01 Order 1
    assets:cash  € -100
    assets:stocks:20200101  1 AAPL @@ € 100

2020-01-02 Order 2
    assets:cash  € -125
    assets:stocks:20200102  1 AAPL @@ € 125

P 2020-01-03 AAPL € 150
```

Using `hledger bs -f test.journal --infer-value -V -D -e 2020-01-04` gives the following output:
```
Balance Sheet 2020-01-01..2020-01-03

                        || 2020-01-01  2020-01-02  2020-01-03 
========================++====================================
 Assets                 ||                                    
------------------------++------------------------------------
 assets:cash            ||     € -100      € -225      € -225 
 assets:stocks:20200101 ||      € 100       € 125       € 150 
 assets:stocks:20200102 ||          0       € 125       € 150 
------------------------++------------------------------------
                        ||          0        € 25        € 75 
========================++====================================
 Liabilities            ||                                    
------------------------++------------------------------------
------------------------++------------------------------------
                        ||                                    
========================++====================================
 Net:                   ||          0        € 25        € 75 
```

Using `hledger bs -f test.journal --infer-value -B -D -e 2020-01-04` gives the following output:
```
Balance Sheet 2020-01-01..2020-01-03, valued at cost

                        || 2020-01-01  2020-01-02  2020-01-03 
========================++====================================
 Assets                 ||                                    
------------------------++------------------------------------
 assets:cash            ||     € -100      € -225      € -225 
 assets:stocks:20200101 ||      € 100       € 100       € 100 
 assets:stocks:20200102 ||          0       € 125       € 125 
------------------------++------------------------------------
                        ||          0           0           0 
========================++====================================
 Liabilities            ||                                    
------------------------++------------------------------------
------------------------++------------------------------------
                        ||                                    
========================++====================================
 Net:                   ||          0           0           0 
```

I think the difference of these two would be an interesting report (essentially reporting the unrealized gains):
```
Balance Sheet 2020-01-01..2020-01-03

                        || 2020-01-01  2020-01-02  2020-01-03 
========================++====================================
 Assets                 ||                                    
------------------------++------------------------------------
 assets:stocks:20200101 ||          0        € 25        € 50 
 assets:stocks:20200102 ||          0           0        € 25 
------------------------++------------------------------------
                        ||          0        € 25        € 75 
========================++====================================
 Liabilities            ||                                    
------------------------++------------------------------------
------------------------++------------------------------------
                        ||                                    
========================++====================================
 Net:                   ||          0        € 25        € 75 
 ```

The UX for this could be a new flag similar to `-V` and `-B` (`-G` or `--gain` in full, for example). The `--value` flag should probably also be extended, perhaps as the following: `--value=gain,COMM`.

**Comment by charvp** [comment]
I have a patch with a basic implementation of what I want: https://ghostbin.co/paste/fv78e/raw
There's still some weird bugs though: the output without `--infer-value` is wrong for the example above, and `bal` doesn't report anything while `bs` does. If you think the patch is somewhat close to something that you would merge I'd be willing to make a pull request and do the work required to get it merged.

**Comment by simonmichael** [comment]
@charvp, nice! I'll look more at this soon I hope. Help from other testers welcome. 

I was thinking it might be a new balance report mode (added to --change, --cumulative, --historical) but a new valuation mode makes more sense I expect.

**Comment by simonmichael** [comment]
If you have time, maybe you could draft the new doc - one or more edits that would go under https://hledger.org/hledger.html#valuation (pasting here is fine).

**Comment by charvp** [comment]
From my usage of the patch I wrote I think a valuation mode makes the most sense, since --change, --cumulative and --historical can all apply when calculating gain. The only thing I wasn't sure about is whether it should be a valuation modifier or a new valuation mode. The first would be more technically correct, I guess, but the second was a lot easier to implement...

**Comment by charvp** [comment]
> If you have time, maybe you could draft the new doc - one or more edits that would go under https://hledger.org/hledger.html#valuation (pasting here is fine).

I'll get to that in the morning.

**Comment by simonmichael** [comment]
PS and @charvp, feel free to make a PR when ready.

**Comment by the-solipsist** [comment]
Hi, @charvp.  I'm thinking of two other cases, and wonder what your opinion is on how they fit within what you're working on.

1. Transactions costs that need to be accounted for when calculating gains.
2. Gains as calculated using trading accounts.

An example of (1) would be:
```
2020-01-01 Order 1
    assets:cash                       € -102
    assets:stocks:20200101  1 AAPL @@ € 100
    expenses:stockbroker              € 2

2020-01-02 Order 2
    assets:cash                       € -125
    assets:stocks:20200102  1 AAPL @@ € 125
    assets:cash                       € -3
    expenses:stockbroker              € 3

P 2020-01-03 AAPL € 150
```

In this scenario, I would want € 2 and € 3 excluded from my gains, since they were costs incurred during the acquisition of the capital.

An example of (2) would be:
```
2020-01-01 Order 1
    assets:cash                       € -102
    equity:trading:stocks:EUR:AAPL    € 100
    equity:trading:stocks:AAPL:EUR    - 1 AAPL
    assets:stocks:20200101              1 AAPL
    expenses:stockbroker              € 2

2020-01-02 Order 2
    assets:cash                       € -125
    equity:trading:stocks:EUR:AAPL    € 125
    equity:trading:stocks:AAPL:EUR    - 1 AAPL
    assets:stocks:20200102              1 AAPL
    assets:cash                       € -3
    expenses:stockbroker              € 3

P 2020-01-03 AAPL € 150
```
This is a software-independent way of capturing multiple-commodity transactions in double-entry accounting.  I'd normally figure the capital gains by doing:
```
hledger -f stock.j bal equity:trading:stocks -V -e 2020-01-04
              € -300  equity:trading:stocks:AAPL:EUR
               € 225  equity:trading:stocks:EUR:AAPL
--------------------
               € -75
```
But as you see, that gains reported is accurate, but technically wrong since it doesn't account for the associated expense.
```
hledger -f stock.j bal assets -V -e 2020-01-04
              € -230  assets:cash
               € 150  assets:stocks:20200101
               € 150  assets:stocks:20200102
--------------------
                € 70

```
This balance query gives me the figure of € 70, taking into account the expenses incurred.  

(However, such a query is more difficult to run on real-world data without accurate lots data (e.g., `assets:cash:20200101`), whereas the balance query for the trading account is easier, as is adding lots data (e.g., `equity:trading:stocks:AAPL:EUR:20200101`) since I usually use trading accounts only in situations with lots, whereas `assets:cash` is used in many situations without lots as well, and so having "lots" sub-accounts for that is more inconvenient.)

At any rate, do you think the `--gains` command you're working on could possibly work with these two situations as well?

**Comment by charvp** [comment]
(1) is very difficult to support (I think), since we would have to look at other postings in the transaction and somehow figure out which posting is for the costs. I solve this in my journal by writing this:
```
2020-01-01 Order 1
    assets:cash                       € -102
    assets:stocks:20200101  1 AAPL @@ € 102  ; € 2 costs

P 2020-01-01 AAPL € 100
```
Not perfect, since I can't easily query for costs, but I'm not very interested in that anyway. Including the costs in the total price also makes sense intuitively to me, since that's total price of what I paid to buy the stocks (which is what @@ means).

(2) Is actually "too balanced" for my implementation. AFAICT, hledger doesn't see any currency conversion, so there isn't a cost basis. The calculation in the patch I wrote uses that cost basis to figure out the gain, so I don't really see how to support this scenario. (ledger's `--gain` also doesn't seem to give any sensible output for it.)

**Comment by simonmichael** [comment]
> I think a valuation mode makes the most sense, since --change, --cumulative and --historical can all apply when calculating gain.

@charvp, could you explain a little more what those combinations would mean ?

> hledger doesn't see any currency conversion, so there isn't a cost basis.

Correct, hledger doesn't know how to convert to cost unless it sees the @ or @@ notation. We'd like to make it infer the cost in this situation, cf [#1177](https://github.com/simonmichael/hledger/issues/1177#issuecomment-648215555).

**Comment by simonmichael** [comment]
Ah, I see. As with --change we also want to see the in-report totals (--cumulative) and all-time totals (--historical), so with --gain (another kind of change) we might want to see the in-report or all-time total gain. 

Another possibility is to create new modes for that (--cumulative-gain, --historical-gain ?). That would allow choosing different valuation types with gain reports, like the other reports.

(bal --gain --value=now, gain until today... 
bal --gain --value=then, gain at time of transaction (difference of cost and market value)... 
bal --gain --value=DATE, gain as of a certain date...)

**Comment by adept** [comment]
`roi` in master includes the code that prices transactions according to `--cost`/`-V` etc. I am not sure if the changes made it into 1.20.4 (probably not)

**Comment by aragaer** [comment]
```
P 2020-01-01 A 9 

2020-01-01 buy
 assets  1A @@ 10
 assets

P 2021-01-01 A 11
```

What command should I use (on master) to see 10% roi for this file?

**Comment by simonmichael** [comment]
> I am not sure if the changes made it into 1.20.4 (probably not)

Correct, they'll be in 1.21 (due march 1).

**Comment by adept** [comment]
@aragaer :

```
$  cat /tmp/1.journal                                                                                                                                           ~
P 2020-01-01 A 9 

2020-01-01 buy
 assets:inv  1A @@ 10
 assets::cash

P 2021-01-01 A 11

$ hledger roi -f /tmp/1.journal --inv inv --pnl pn -Y -V                                                                                                       ~
+---++------------+------------++---------------+----------+-------------+-----++-------+-------+
|   ||      Begin |        End || Value (begin) | Cashflow | Value (end) | PnL ||   IRR |   TWR |
+===++============+============++===============+==========+=============+=====++=======+=======+
| 1 || 2020-01-01 | 2020-12-31 ||             0 |       10 |          11 |   1 || 9.97% | 9.97% |
+---++------------+------------++---------------+----------+-------------+-----++-------+-------+

```

**Comment by aragaer** [comment]
Great. Next step would be to specify actual commodity somewhere in --inv so that I could track it without creating per-commodity accounts.

**Comment by adept** [comment]
--inv is a query, so presumably you should be able to do this already. I
don't have docs in front of me, but I vaguely remember that you can specify
commodity in queries

**Comment by simonmichael** [comment]
`cur:REGEX`, https://hledger.org/hledger.html#queries

**Comment by aragaer** [comment]
Yes but...

I have 2 metals - gold (cur:XAU) and silver (cur:XAG). Both are in the same account "metal".

- `stack exec -- hledger roi --inv 'metal' --pnl pn -V` works fine
- `stack exec -- hledger roi --inv 'cur:XAU' --pnl pn -V` results in "error: Not bracketed"
- same for `cur:XAG` and even `cur:XA.` (which should match both).

I might try and create a journal that reproduces the issue but for now I feel that --inv "cur:XA." is not the same as --inv "metal" for me.

**Comment by aragaer** [comment]
Here's the journal: I've bought 
```
P 2020-10-02 A 48

2020-10-02
 assets:inv  1 A @ 50
 assets

P 2020-10-16 A 49
```

```
$ stack exec -- hledger -f a.journal roi --inv inv --pnl pn -V
+---++------------+------------++---------------+----------+-------------+-----++---------+---------+
|   ||      Begin |        End || Value (begin) | Cashflow | Value (end) | PnL ||     IRR |     TWR |
+===++============+============++===============+==========+=============+=====++=========+=========+
| 1 || 2020-10-02 | 2020-10-02 ||             0 |       50 |          49 |  -1 || -99.94% | -99.94% |
+---++------------+------------++---------------+----------+-------------+-----++---------+---------+
$ stack exec -- hledger -f a.journal roi --inv cur:A --pnl pn -V
hledger: Error (NotBracketed): No solution for Internal Rate of Return (IRR).
  Possible causes: IRR is huge (>1000000%), balance of investment becomes negative at some point in time.
```

But maybe this should be a separate issue.

**Comment by adept** [comment]
I have something to say on both roi issues raised in recent comments, but i propose to stop hijacking this ticket and open separate ones instead

**Comment by Xitian9** [comment]
Now that the `--valuechange` functionality has been merged, getting this `--gain` behaviour merged is the next thing in line. 

It seems that a UI has been decided on in https://github.com/simonmichael/hledger/issues/1353#issuecomment-772594810, but I'm not certain that the proper behaviour has been nailed down yet. At the very least, I don't understand it.

Questions that I have:
1. Does `--gain` just report on the difference in valuation of price between purchase time and now/period end/whatever, or does it use the cost at purchase time. Concretely, what would you expect it to output given the following journal:
```
P 2020-01-01 A B2
P 2020-02-01 A B3

2020-01-01 Buy it
    assets    1A @ 1B
    expenses

2020-01-01 Buy something else
    assets    10A
    expenses
```
Which result would you expect?
```
$ hledger bal --gain --periodic -M assets

        || Jan  Feb
========++==========
 assets ||   0   B1
--------++----------
        ||   0   B1
```
or
```
$ hledger bal --gain --periodic -M assets

        || Jan  Feb
========++==========
 assets ||  B1   B1
--------++----------
        ||  B1   B1
```
or something else entirely?

2. What should be done regarding the points raised by @the-solipsist here: https://github.com/simonmichael/hledger/pull/1432#issuecomment-760117089? I don't understand all the acronyms, but it gives the impression that figuring out what to present is more complicated than what we're considering here, or at least it can be.

**Comment by the-solipsist** [comment]
My comment on that pull request both brought up a substantive issue (what
the cost basis is for calculating gains) as well as an issue of form.

The substantive issue of cost basis (whether first-in-first-out, average
cost basis, etc.) can be kicked down the road, since it requires
lot-tracking functionality, which hledger does not currently have.

But the issue I raised about form (whether it should be `--value=gain`)
still remains. It might even be due to a misunderstanding on my part as to
what exactly this function does or is supposed to do. But it would be good,
in any case, to know.

**Comment by chvp** [comment]
1. The cost at purchase time is what I would use. The valuation at the time of buying will usually be lower than cost basis, and the difference can immediately represent extra costs.
2. I think most things talked about in that comment are things solved by the proposed UI in https://github.com/simonmichael/hledger/issues/1353#issuecomment-772594810. The cost basis issue they touch upon is somewhat orthogonal I think. Since what I propose would use the cost basis in the postings, I would think it would be up to the accountant to track it correctly in their journal.

**Comment by the-solipsist** [comment]
I'm sorry, I don't think I was very clear.  

When I said, "what the cost basis is for calculating gain," I was assuming _multiple_ purchases of the same commodity in the same account, and not just a single purchase.  In case of multiple transactions (say, where you've bought a units of a share at different prices on different dates, and are then selling them), the question arises as to how to calculate the cost basis, since the gain depends on how the cost basis is calculated.  Which method one may use is often prescribed by your country's tax laws, and usually also depends on what kind of asset it is.  My understanding is the first-in-first-out (FIFO) is the most common cost basis.  In some countries, like Canada, an average cost basis (ACB) is used.  Under most tax laws, an adjusted basis (i.e., adjusted to account for improvements, commissions & fees, and even inflation in some jurisdictions, etc.) is used.

Here's [an example from Investopedia which demonstrates the different gains from different cost bases] (https://www.investopedia.com/terms/c/costbasis.asp#:~:text=Cost%20basis%20is%20the%20original,and%20the%20current%20market%20value.):

> Cost Basis Comparisons
> Cost basis comparison can be an important consideration. Assume that an investor made the following consecutive fund purchases in a taxable account: 1,500 shares at $20, 1,000 shares at $10 and 1,250 shares at $8. The investor’s average cost basis is calculated by dividing $50,000/3,750 shares. The average cost is $13.33.
> 
> Suppose the investor then sells 1,000 shares of the fund at $19. The investor would have a capital gain of $5,670 using the average cost basis method.
> 
> Gain/loss using average cost basis: ($19 - $13.33) x 1,000 shares = $5,670
> Results can vary significantly by cost basis.
> 
> First in first out: ($19 - $20) x 1,000 shares = - $1,000
> Last in first out: ($19 - $8) x 1,000 = $11,000
> High cost: ($19 - $20) x 1,000 shares = - $1,000
> Low cost: ($19 - $8) x 1,000 = $11,000
> In this case, the investor would be better off if they had selected the FIFO method or the high cost method to determine the cost basis prior to selling the shares. These methods would result in no tax on a loss of $1,000. With the average cost basis method, the investor must pay a capital gains tax on the gain of $5,670.

Beancount calls this "[booking method](https://beancount.github.io/docs/how_inventories_work.html)", and allows for it to be set using `option "booking_method" "FIFO"` or on a per-account basis using `2016-05-01 open Assets:Vanguard:RGAGX  "AVERAGE"`.

**Comment by chvp** [comment]
That is how I understood it. I think in the current support of lots (which is to say, none) I think it would make sense to leave this to the accountant themselves. FIFO would require no real extra work, but others would be more involved (essentially clearing the amounts in the account and adding them again using the correct cost basis).

**Comment by the-solipsist** [comment]
So, in the Investopedia example, assuming that instead of "the investor then sells 1,000 shares of the fund at $19", it is "the market price of the fund becomes $19", what amount would the `--gain` function show?  (So, instead of a sale being recorded in the journal, the price `P 2021-03-02 shares $19` is recorded, and then the `--gain` function is run to cover the entire period from the first transaction.)

**Comment by Xitian9** [comment]
Thank you @the-solipsist, that makes things very clear.

**Comment by Xitian9** [comment]
Can this be closed with the merging of #1623?

**Comment by chvp** [comment]
I think so, yes.
